lib/md051.js

```diff
@@ -119,12 +147,27 @@ module.exports = {
         for (const definition of definitions) {
           const { endColumn, startColumn } = definition;
           const text = unescapeStringTokenText(definition);
+          const textFragment = `#${encodeURIComponent(text.slice(1))}`
           if (
             (text.length > 1) &&
             text.startsWith("#") &&
-            !fragments.has(text) &&
-            !fragments.has(`#${encodeURIComponent(text.slice(1))}`)
+            !fragments.has(textFragment)
           ) {
+            if (textFragment.startsWith("#L")) {
```

Comment:
This `if` and everything in it feels like it could be replaced with something like:

```
const lineFragmentRe = /^#L\d+$/;
```

And then a single addition to the `if` checks above:

```
!lineFragmentRe.test(textFragment)
```

And remove the two functions added higher up in this file.

Reply:
This doesn't work, as `!lineFragmentRe.test(textFragment)` returns false with `#L12-abc`, but it could be a valid heading fragment (there is a test covering that).

However, we can still use this Regex, and replace the "2 functions" (`getLineNumberStringFromFragment` and `isValidIntegerString`), but we need to make sure that we ignore all `encodedText` matching `/^#L\d+$/` before verifying in `fragments` Map (still with a `continue`, I don't know if there's another easy alternative to do it), and then, we need to lower case `encodedText` to check if it is in `fragments` as the Map contains `#l12-abc`, not `#L12-abc`.

Talking about lowercase, I don't understand that test case:
```md
## Valid Fragments

[Title](#Valid-Fragments) {MD051}
```

`#Valid-Fragments` is valid, as GitHub doesn't care about case sensitivity for Heading Fragments, same validity as  `[Valid](#valid-fragments)`, so in this case the rule should not returns an error.
Is it for consistency that, the rule throws an error here?

Relaxing that part, could avoid that conditon:
```js
if (encodedText.startsWith('#L')) {
  encodedText = encodedText.toLowerCase();
}
```

And, instead of `inlineText.toLowerCase()` in `convertHeadingToHTMLFragment`, we could `.toLowerCase()` the whole return value.

Reply:
While there are multiple forms that work in practice, GitHub produces links for headings that are all lower-case, so that's what this rule wants to see.